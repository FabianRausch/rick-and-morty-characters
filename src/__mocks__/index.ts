export * from "./getAllCharactersMocks";
export * from "./getMultipleEpisodesMocks";
export * from "./charactersMock";
export * from "./episodesMock";
export * from "./longIdsMock";
