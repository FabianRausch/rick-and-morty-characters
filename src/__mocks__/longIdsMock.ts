export const longIds = [
  "https://rickandmortyapi.com/api/character/1",
  "https://rickandmortyapi.com/api/character/2",
];
