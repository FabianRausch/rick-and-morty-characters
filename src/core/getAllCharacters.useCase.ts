import type { AllCharactersData } from "@/interfaces";
import { isPositiveInteger } from "@/helpers";
import { charactersMapper } from "@/mappers";

interface Props {
  page?: number | string;
}

const API_URL = process.env.NEXT_PUBLIC_RICK_AND_MORTY_API;

export const getAllCharactersUseCase = async ({
  page,
}: Props): Promise<AllCharactersData> => {
  try {
    const isValidPage = isPositiveInteger(page as string);
    const urlParams = new URLSearchParams();
    if (isValidPage) urlParams.append("page", page as string);

    const resp = await fetch(`${API_URL}/character?${urlParams.toString()}`, {
      next: { revalidate: 3600 },
    });

    if (!resp.ok) throw new Error(`Error on service ==> ${resp.url}`);

    const data: AllCharactersData = await resp.json();
    return charactersMapper(data);
  } catch (error) {
    console.log("Error on getAllCharactersUseCase =>", error);
    throw new Error("Something went wrong");
  }
};
