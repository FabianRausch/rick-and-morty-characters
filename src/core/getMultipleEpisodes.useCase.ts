import { SingleEpisode } from "@/interfaces";
import { isPositiveInteger } from "@/helpers";
import { episodesMapper } from "@/mappers";

interface Props {
  episodeIds: string[];
}

const API_URL = process.env.NEXT_PUBLIC_RICK_AND_MORTY_API;

export const getMultipleEpisodesUseCase = async ({
  episodeIds,
}: Props): Promise<SingleEpisode[]> => {
  try {
    const uniqueIds = [...new Set(episodeIds)];
    const validIds = uniqueIds.filter((episodeId) =>
      isPositiveInteger(episodeId)
    );
    if (!validIds.length)
      throw new Error(
        `Error on getMultipleEpisodesUseCase episodesId path is empty `
      );

    const resp = await fetch(`${API_URL}/episode/${validIds.toString()}`, {
      next: { revalidate: 3600 },
    });

    if (!resp.ok) throw new Error(`Error on service ==> ${resp.url}`);

    const data: SingleEpisode[] = await resp.json();
    return episodesMapper(data);
  } catch (error) {
    console.log("Error on getMultipleEpisodesUseCase =>", error);
    throw new Error("Something went wrong");
  }
};
