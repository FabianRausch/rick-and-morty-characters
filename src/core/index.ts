export * from "./getAllCharacters.useCase";
export * from "./getMultipleEpisodes.useCase";
