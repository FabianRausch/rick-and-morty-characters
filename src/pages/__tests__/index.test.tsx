import { render } from "@testing-library/react";
import Component, { getServerSideProps } from "..";
import { mappedCharactersApiResponse } from "@/__mocks__";

jest.mock("@/core", () => ({
  getAllCharactersUseCase: jest.fn(),
}));

import { getAllCharactersUseCase } from "@/core";

const mockGetAllCharactersUseCase = getAllCharactersUseCase as jest.Mock;

describe("Test on Home page", () => {
  it("Should match with snapshot", () => {
    const { asFragment } = render(
      <Component data={mappedCharactersApiResponse} />
    );
    expect(asFragment).toMatchSnapshot();
  });
});

describe("Test on getServerSideProps Home", () => {
  beforeEach(() => {
    mockGetAllCharactersUseCase.mockClear();
  });

  it("should return initial characters data on successful fetch", async () => {
    mockGetAllCharactersUseCase.mockResolvedValueOnce(
      mappedCharactersApiResponse
    );

    const result = await getServerSideProps();
    expect(mockGetAllCharactersUseCase).toHaveBeenCalledTimes(1);
    expect(result).toEqual({
      props: {
        data: mappedCharactersApiResponse,
      },
    });
  });

  it("should return error on fetch", async () => {
    mockGetAllCharactersUseCase.mockRejectedValueOnce(
      mappedCharactersApiResponse
    );

    const result = await getServerSideProps();
    expect(mockGetAllCharactersUseCase).toHaveBeenCalledTimes(1);
    expect(result).toEqual({
      redirect: {
        destination: "/error",
        permanent: false,
      },
    });
  });
});
