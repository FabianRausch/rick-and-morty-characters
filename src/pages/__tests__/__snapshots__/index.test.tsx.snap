// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Test on Home page Should match with snapshot 1`] = `[Function]`;
